serializers/v0x04.py

```diff
@@ -0,0 +1,140 @@
+            'dl_type': OxmOfbMatchField.OFPXMT_OFB_ETH_TYPE,
+            'dl_vlan': OxmOfbMatchField.OFPXMT_OFB_VLAN_VID,
+            'dl_vlan_pcp': OxmOfbMatchField.OFPXMT_OFB_VLAN_PCP}
+        # Invert match_values index
+        self._match_values = {b: a for a, b in self._match_names.items()}
+
+    def from_dict(self, dictionary):
+        """Return an OF 1.0 FlowMod message from serialized dictionary."""
+        flow_mod = FlowMod()
+        instruction = InstructionApplyAction()
+        flow_mod.instructions.append(instruction)
+
+        for field, data in dictionary.items():
+            if field in self.flow_attributes:
+                setattr(flow_mod, field, data)
+            elif field == 'match':
+                tlvs = self._match_from_dict(data)
+                flow_mod.match.oxm_match_fields.append(list(tlvs))
+            elif field == 'actions':
+                actions = self._actions_from_dict(data)
+                instruction.actions.extend(list(actions))
+
+        return flow_mod
+
+    def _match_from_dict(self, dictionary):
+        known_fields = ((field, data) for field, data in dictionary.items()
+                        if field in self._match_names)
+        for field_name, data in known_fields:
+            tlv = OxmTLV()
+            tlv.oxm_field = self._match_names[field_name]
+            # set oxm_value
+            if field_name == 'dl_vlan_pcp':
+                tlv.oxm_value = data.to_bytes(1, 'big')
+            elif field_name == 'dl_vlan':
+                vid = data | VlanId.OFPVID_PRESENT
+                tlv.oxm_value = vid.to_bytes(2, 'big')
+            elif field_name in ('dl_src', 'dl_dst'):
+                tlv.oxm_value = HWAddress(data).pack()
+            elif field_name == 'in_port':
+                tlv.oxm_value = data.to_bytes(4, 'big')
+            else:
+                tlv.oxm_value = data.to_bytes(2, 'big')
+            yield tlv
+
+    @classmethod
+    def _actions_from_dict(cls, dictionary):
+        for action_type, data in dictionary.items():
+            action = cls._action_from_dict(action_type, data)
+            if action:
+                yield action
+
+    @classmethod
+    def _action_from_dict(cls, action_type, data):
+        if action_type == 'set_vlan':
+            tlv = cls._create_vlan_tlv(vlan_id=data)
+            return ActionSetField(field=tlv)
+        elif action_type == 'output':
+            return ActionOutput(port=data)
+
+    @staticmethod
+    def _create_vlan_tlv(vlan_id):
+        tlv = OxmTLV()
+        tlv.oxm_field = OxmOfbMatchField.OFPXMT_OFB_VLAN_VID
+        oxm_value = vlan_id | VlanId.OFPVID_PRESENT
+        tlv.oxm_value = oxm_value.to_bytes(2, 'big')
+        return tlv
+
+    def to_dict(self, flow_stats):
+        """Return a dictionary created from input 1.3 switch's flows."""
+        flow_dict = {field: data.value
+                     for field, data in vars(flow_stats).items()}
```

Comment:
Here we need to specify which items we want: `field` must be in `flow_attributes`.

I'm getting: 
`Match object has no attribute 'value'`

Reply:
(same for v0x01, in line 58)